dll/win32/shell32/shlexec.cpp

```diff
@@ -1431,60 +1428,139 @@ static HRESULT shellex_run_context_menu_default(IShellExtInit *obj,
+        HRESULT hres;
+    };
+}
 
+static HRESULT shellex_load_object_and_run(HKEY hkey, LPCGUID guid, LPSHELLEXECUTEINFOW sei)
+{
     TRACE("%p %s %p\n", hkey, debugstr_guid(guid), sei);
 
-    r = CoInitialize(NULL);
-    if (FAILED(r))
-        goto end;
+    CCoInit coInit;
+
+    if (FAILED_UNEXPECTEDLY(coInit.hres))
+        return coInit.hres;
 
-    r = CoCreateInstance(*guid, NULL, CLSCTX_INPROC_SERVER,
+    CComPtr<IShellExtInit> obj;
+    HRESULT hr = CoCreateInstance(*guid, NULL, CLSCTX_INPROC_SERVER,
                          IID_PPV_ARG(IShellExtInit, &obj));
-    if (FAILED(r))
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    CComPtr<IDataObject> dataobj;
+    hr = shellex_get_dataobj(sei, dataobj);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    hr = obj->Initialize(NULL, dataobj, hkey);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    CComPtr<IObjectWithSite> ows;
+    hr = obj->QueryInterface(IID_PPV_ARG(IObjectWithSite, &ows));
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    ows->SetSite(NULL);
+
+    return shellex_run_context_menu_default(obj, sei);
+}
+
+static HRESULT shellex_get_contextmenu(LPSHELLEXECUTEINFOW sei, CComPtr<IContextMenu>& cm)
+{
+    CComHeapPtr<ITEMIDLIST> allocatedPidl;
+    LPITEMIDLIST pidl = NULL;
+
+    if (sei->lpIDList)
     {
-        ERR("failed %08x\n", r);
-        goto end;
+        pidl = (LPITEMIDLIST)sei->lpIDList;
     }
-
-    dataobj = shellex_get_dataobj(sei);
-    if (!dataobj)
+    else
     {
-        ERR("failed to get data object\n");
-        r = E_FAIL;
-        goto end;
+        SFGAOF sfga = 0;
+        HRESULT hr = SHParseDisplayName(sei->lpFile, NULL, &allocatedPidl, SFGAO_STORAGECAPMASK, &sfga);
+        if (FAILED(hr))
+        {
+            WCHAR Buffer[MAX_PATH] = {};
+            // FIXME: MAX_PATH.....
+            UINT retval = SHELL_FindExecutable(sei->lpDirectory, sei->lpFile, sei->lpVerb, Buffer, _countof(Buffer), NULL, NULL, NULL, sei->lpParameters);
+            if (retval <= 32)
+                return HRESULT_FROM_WIN32(retval);
+
+            hr = SHParseDisplayName(Buffer, NULL, &allocatedPidl, SFGAO_STORAGECAPMASK, &sfga);
+            // This should not happen, we found it...
+            if (FAILED_UNEXPECTEDLY(hr))
+                return hr;
+        }
+
+        pidl = allocatedPidl;
     }
 
-    r = obj->Initialize(NULL, dataobj, hkey);
-    if (FAILED(r))
-        goto end;
+    CComPtr<IShellFolder> shf;
+    LPCITEMIDLIST pidllast = NULL;
+    HRESULT hr = SHBindToParent(pidl, IID_PPV_ARG(IShellFolder, &shf), &pidllast);
+    if (FAILED(hr))
+        return hr;
 
-    r = obj->QueryInterface(IID_PPV_ARG(IObjectWithSite, &ows));
-    if (FAILED(r))
-        goto end;
+    return shf->GetUIObjectOf(NULL, 1, &pidllast, IID_NULL_PPV_ARG(IContextMenu, &cm));
+}
 
-    ows->SetSite(NULL);
+static HRESULT IContextMenu_exec(LPSHELLEXECUTEINFOW sei)
```

Comment:
Perhaps a different name could be chosen here. We already have a different code path that invokes the IContextMenu of shell extensions and this could become confusing.

Reply:
Thanks, will do